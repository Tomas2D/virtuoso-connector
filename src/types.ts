import type { SparqlQuery } from '@tpluscode/sparql-builder/lib';
import type { TemplateResult } from '@tpluscode/rdf-string/lib/TemplateResult';

export interface IDatabaseConnection {
  rawQuery<T extends any = any>(query: string, procedureCalls: boolean): Promise<T>;
  query<T extends any = any>(
    query: string | SparqlQuery | TemplateResult<unknown>,
    checkVoidExpression: boolean,
  ): Promise<T>;
}

export interface IDatabaseConnectionConfig {
  url: string;
  username: string;
  password: string;
  driverPath: string;
  lazy?: boolean;
  maxQueryTimeout?: number;
  poolSize?: number;
}

// Child process can only retrieve simple literal values
type SerializedTypes<T extends unknown[]> = T extends []
  ? []
  : T extends [infer H, ...infer R]
  ? H extends string | boolean | number | null | undefined
    ? [H, ...SerializedTypes<R>]
    : never
  : T;

export interface IChildTask<K extends keyof IDatabaseConnection = keyof IDatabaseConnection> {
  id: string;
  method: K;
  params: SerializedTypes<Parameters<IDatabaseConnection[K]>>;
}

type _ChildTaskResponse<K extends keyof IDatabaseConnection> = {
  id: string;
  method: K;
};

export type IChildTaskSuccessResponse<
  K extends keyof IDatabaseConnection = keyof IDatabaseConnection,
> = _ChildTaskResponse<K> & {
  result: Awaited<ReturnType<IDatabaseConnection[K]>>;
};

export type IChildTaskErrorResponse<
  K extends keyof IDatabaseConnection = keyof IDatabaseConnection,
> = _ChildTaskResponse<K> & {
  error: string;
};

export type ChildTaskResponse<K extends keyof IDatabaseConnection = keyof IDatabaseConnection> =
  | IChildTaskSuccessResponse<K>
  | IChildTaskErrorResponse<K>;
