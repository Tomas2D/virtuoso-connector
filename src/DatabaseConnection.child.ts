import { JDBC, Java } from '@naxmefy/jdbc';
import { mergeWith, min, noop, omit } from 'lodash';
import { ResultSet, Statement, Connection } from '@naxmefy/jdbc';
import type { IChildTask, ChildTaskResponse, IDatabaseConnection } from './types';
import type { IDatabaseConnectionConfig } from './types';

class DatabaseConnectionChild implements IDatabaseConnection {
  protected readonly db: JDBC;

  async destroy() {
    const connection = await this.db.getConnection(false);
    if (connection) {
      await connection.close();
    }
  }

  constructor(options: IDatabaseConnectionConfig) {
    const { driverPath, lazy = true } = options;

    const jv = new Java(true, false);
    if (!jv.isJvmCreated()) {
      jv.addClasspath([driverPath]);
    }

    const config = mergeWith(
      {
        className: 'virtuoso.jdbc4.Driver',
        maxidle: 24 * 60 * 60 * 1000,
        properties: {},
        url: 'URL_MISSING',
      },
      omit<IDatabaseConnectionConfig>(options, 'driverPath', 'lazy', 'maxQueryTimeout'),
      <T>(objValue: T, srcValue: T) => {
        if (srcValue !== undefined) {
          return srcValue;
        }
        return objValue;
      },
    );

    this.db = new JDBC(config);
    if (lazy !== true) {
      this.db.getConnection(true).then(noop);
    }
  }

  protected isQueryWithResult(query: string): boolean {
    const countIndex = query.match(/COUNT/i)?.index ?? -1;
    const selectIndex = query.match(/SELECT/i)?.index ?? -1;
    const constructIndex = query.match(/CONSTRUCT/i)?.index ?? -1;
    const askIndex = query.match(/CONSTRUCT/i)?.index ?? -1;

    const sMin =
      min([countIndex, selectIndex, constructIndex, askIndex].filter((i) => i !== -1)) ?? Infinity;
    if (sMin === 0) {
      return true;
    }

    const deleteIndex = query.match(/DELETE/i)?.index ?? -1;
    const insertIndex = query.match(/INSERT/i)?.index ?? -1;
    const clearIndex = query.match(/INSERT/i)?.index ?? -1;
    const rMin = min([deleteIndex, insertIndex, clearIndex].filter((i) => i !== -1)) ?? Infinity;

    return sMin < rMin;
  }

  protected async execute(
    query: string,
    statement: Statement,
    connection: Connection,
    checkVoidExpression: boolean,
  ) {
    const isSelect = this.isQueryWithResult(query);

    const prefix = [`SPARQL define output:format "JSON"`];
    if (checkVoidExpression) {
      prefix.push(`define sql:signal-void-variables 1`);
    }

    const fullQuery = prefix.join(' ').concat(query).concat('\n');

    const resultSet = await statement.executeQuery(fullQuery).catch((err: Error) => {
      connection.close();
      throw err;
    });

    if (!isSelect) {
      return null;
    }

    const results = await resultSet.fetchAllResults();
    return results[0]['callret-0'];
  }

  async rawQuery<T>(query: string, procedureCalls = false): Promise<T> {
    const connection = await this.db.getConnection(true);

    const statement = await connection.createStatement();
    const processor = statement[procedureCalls ? 'executeUpdate' : 'executeQuery'];

    const rawResults: ResultSet | number = await processor(query).catch((err: Error) => {
      connection.close();
      throw err;
    });

    if (typeof rawResults !== 'object') {
      return rawResults as unknown as T;
    }
    return rawResults.fetchAllResults() as unknown as Promise<T>;
  }

  async query<T>(queryString: string, checkVoidExpression = true): Promise<T> {
    console.info(`Waiting for connection...`);
    const connection = await this.db.getConnection(true);
    console.info(`Creating statement...`);
    const statement = await connection.createStatement();

    console.info(`Executing...`);
    return this.execute(queryString, statement, connection, checkVoidExpression);
  }
}

// Initialize database connection
const options = JSON.parse(process.argv[2]) as IDatabaseConnectionConfig;
const db = new DatabaseConnectionChild(options);

// Helper function for sending responses to parent process
function respondWithMessage(message: ChildTaskResponse) {
  if (!process.connected) {
    return;
  }
  process.send(message);
}

process.on('message', function (message: IChildTask) {
  log(`Processing task ${message.id}`);
  db[message.method]
    .call(db, ...message.params)
    .then((response) => {
      respondWithMessage({
        id: message.id,
        method: message.method,
        result: response,
      });
    })
    .catch((err) => {
      console.error(err);
      respondWithMessage({
        id: message.id,
        method: message.method,
        error: typeof err === 'string' ? err : String(err),
      });
    })
    .finally(() => {
      console.info(`Processing of task with ID: ${message.id} done!`);
    });
});

function log(...params: unknown[]) {
  if (process.env.NODE_ENV !== 'test') {
    console.debug(...params);
  }
}

function terminate() {
  log(`${DatabaseConnectionChild.name}: Shutdown the child process...`);
  db.destroy().finally(() => {
    log(`${DatabaseConnectionChild.name}: Child process has been disconnected`);
    process.exit(0);
  });
}

process.once('disconnect', terminate);
process.once('sigint', terminate);
process.once('sigterm', terminate);
process.once('sigkill', terminate);
