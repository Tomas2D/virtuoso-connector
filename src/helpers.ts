import type { SparqlQuery } from '@tpluscode/sparql-builder/lib';
import { TemplateResult } from '@tpluscode/rdf-string/lib/TemplateResult';
import { IChildTaskSuccessResponse } from './types';
import VirtuosoResultParser from './utils/VirtuosoResult.parser';

export function queryToString(query: string | SparqlQuery | TemplateResult<unknown>) {
  if (typeof query === 'string') {
    return query;
  }
  if ('build' in (query as object)) {
    return (query as SparqlQuery).build();
  }
  if ('toString' in (query as object)) {
    return query.toString();
  }

  throw new Error(`"query" parameter is invalid type`);
}

export function responseToTriples(message: IChildTaskSuccessResponse) {
  const parser = new VirtuosoResultParser();
  parser.write(message.result);
  parser.end();

  return new Promise((resolve, reject) => {
    const parts: unknown[] = [];
    parser
      .on('data', (data) => parts.push(data))
      .on('error', reject)
      .on('end', () => resolve(parts));
  });
}
