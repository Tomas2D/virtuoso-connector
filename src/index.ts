export * from './DatabaseConnection';
export * from './types';
