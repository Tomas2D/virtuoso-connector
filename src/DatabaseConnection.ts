import { TaskMap } from 'promise-based-task';
import { ChildTask } from './utils/ChildTask';
import { fork, ChildProcess } from 'child_process';
import * as path from 'path';

import type { SparqlQuery } from '@tpluscode/sparql-builder/lib';
import type { TemplateResult } from '@tpluscode/rdf-string/lib/TemplateResult';
import type {
  IDatabaseConnectionConfig,
  IChildTask,
  IChildTaskErrorResponse,
  ChildTaskResponse,
  IChildTaskSuccessResponse,
  IDatabaseConnection,
} from './types';
import { queryToString, responseToTriples } from './helpers';

export class DatabaseConnection implements IDatabaseConnection {
  // Task related properties
  protected tasks = new TaskMap<string, ChildTask>();
  protected taskToProcess = new WeakMap<ChildProcess, string[]>();
  protected lastTaskId = 0;
  protected readonly maxQueryTimeout: number = 0;

  // Pool related properties
  protected childProcessIndex = 0;
  protected childProcesses: ChildProcess[] = [];
  protected readonly maxPoolSize: number = 1;

  constructor(protected readonly options: IDatabaseConnectionConfig) {
    this.maxQueryTimeout = Math.max(Number(options.maxQueryTimeout) || 0, 0);
    this.maxPoolSize = Math.max(Number(options.poolSize) || 1, 1);
  }

  /**
   * Executes ISQL query or procedure
   * @param query
   * @param procedureCalls
   */
  async rawQuery<T>(query: string, procedureCalls = false): Promise<T> {
    const command: IChildTask = {
      id: String(++this.lastTaskId),
      method: 'rawQuery',
      params: [query, procedureCalls],
    };

    return this.sendCommand<T>(command);
  }

  /**
   * Execute SPARQL query
   * @param query
   * @param checkVoidExpression (Virtuoso internal)
   */
  async query<T extends any = any>(
    query: string | SparqlQuery | TemplateResult<unknown>,
    checkVoidExpression = true,
  ): Promise<T> {
    const queryString = queryToString(query);

    const command: IChildTask = {
      id: String(++this.lastTaskId),
      method: 'query',
      params: [queryString, checkVoidExpression],
    };

    return this.sendCommand<T>(command);
  }

  /**
   * Cancel all running queries and destroy all connections
   * class instance can be re-used later
   */
  destroy() {
    this.childProcesses.forEach((childProcess) => childProcess.disconnect());
    this.childProcesses.length = 0;
    this.childProcessIndex = 0;
    this.tasks.clear();
  }

  protected resolveTask(message: IChildTaskSuccessResponse) {
    this.tasks.get(message.id)?.resolve(message.result);
    this.tasks.delete(message.id);
  }

  protected rejectTask(message: IChildTaskErrorResponse) {
    this.tasks.get(message.id)?.reject(message.error);
    this.tasks.delete(message.id);
  }

  /**
   * Send command to the child process
   * @param command
   * @protected
   */
  protected async sendCommand<T>(command: IChildTask): Promise<T> {
    if (this.lastTaskId >= Number.MAX_SAFE_INTEGER) {
      this.lastTaskId = 0;
    }

    const childProcess = this.getChild();
    childProcess.send(command);

    const task = new ChildTask();
    this.tasks.set(command.id, task);
    if (this.taskToProcess.has(childProcess)) {
      this.taskToProcess.get(childProcess)!.push(command.id);
    } else {
      this.taskToProcess.set(childProcess, [command.id]);
    }

    let timeoutId: NodeJS.Timeout | undefined;
    if (this.maxQueryTimeout > 0) {
      timeoutId = setTimeout(() => {
        task.reject(new Error(`Query timeout has been exceeded!`));
        this.destroyChild(childProcess);
      }, this.maxQueryTimeout * 1000);
    }

    return task.finally(() => clearTimeout(timeoutId)) as Promise<T>;
  }

  /**
   * Creates child process, adds it to the pool and bind event handlers
   */
  protected initChild(): void {
    const childProcess = fork(
      path.join(__dirname, './DatabaseConnection.child' + path.extname(__filename)),
      [JSON.stringify(this.options)],
      {
        detached: true,
      },
    );

    childProcess.once('error', () => {
      this.destroyChild(childProcess);
    });

    childProcess.on('message', (message: ChildTaskResponse) => {
      if ('error' in message) {
        this.rejectTask(message);

        if (message.error.includes('Connection to the server lost')) {
          this.destroy();
          return;
        }

        if (
          message.error.includes('Connection failed') ||
          message.error.includes('Broken pipe') ||
          message.error.includes('Connection refused')
        ) {
          this.destroyChild(childProcess);
        }
        return;
      }

      if (message.result === null || message.result === undefined) {
        return this.resolveTask(message);
      }

      responseToTriples(message).then((result: typeof message['result']) => {
        this.resolveTask({
          ...message,
          result,
        });
      });
    });

    this.childProcesses.push(childProcess);
  }

  /**
   * Retrieve child process from the pool (Round Robin Load Balancing)
   * @protected
   */
  protected getChild(): ChildProcess {
    if (this.childProcesses.length < this.maxPoolSize) {
      this.initChild();
      this.childProcessIndex++;
      return this.childProcesses[this.childProcesses.length - 1];
    }

    if (this.childProcessIndex >= Number.MAX_SAFE_INTEGER) {
      this.childProcessIndex = 0;
    }

    const child = this.childProcesses[this.childProcessIndex % this.childProcesses.length];
    this.childProcessIndex++;
    return child;
  }

  /**
   * Disconnect and remove the child process from the pool
   * @param childProcess
   * @protected
   */
  protected destroyChild(childProcess: ChildProcess): void {
    const index = this.childProcesses.findIndex((el) => el === childProcess);
    if (index > -1) {
      if (this.taskToProcess.has(childProcess)) {
        this.taskToProcess.get(childProcess)!.forEach((id) => {
          this.tasks.delete(id);
        });
        this.taskToProcess.get(childProcess)!.length = 0;
      }
      this.childProcesses.splice(index, 1);
    }
    if (childProcess && childProcess.connected) {
      childProcess.disconnect();
    }
  }
}

export default DatabaseConnection;
