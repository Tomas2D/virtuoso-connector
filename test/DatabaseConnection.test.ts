import { DatabaseConnection } from '../src';

describe('Database connection with queries', () => {
  let db: DatabaseConnection;
  const TEST_GRAPH_IRI = 'http://github.com/Tomas2D/virtuoso-connector';

  beforeAll(async () => {
    db = new DatabaseConnection({
      driverPath: process.env.DB_DRIVER_PATH,
      url: process.env.DB_URL,
      username: process.env.DB_USER,
      password: process.env.DB_PASS,
      lazy: false,
      maxQueryTimeout: 5,
      poolSize: 1,
    });
  });

  afterAll(() => {
    db.destroy();
  });

  it('Verify connection', async () => {
    await expect(db.query(`ASK {}`)).resolves.toBeDefined();
  });

  it('Insert data', async () => {
    await db.query(`
      PREFIX ns: <http://example.org/ns#>
      
      INSERT DATA { 
        GRAPH <${TEST_GRAPH_IRI}> { 
          <http://example/book1> ns:name "Book #1"@en . 
          <http://example/book1> ns:price 100 .
          
          <http://example/book1> ns:name "Book #2"@en . 
          <http://example/book1> ns:price 200 . 
        }
      }
    `);
  });

  it('Retrieves data', async () => {
    const books = await db.query(`
      PREFIX dc: <http://purl.org/dc/elements/1.1/>
      PREFIX ns: <http://example.org/ns#>
      
      SELECT DISTINCT ?name ?price
      FROM <${TEST_GRAPH_IRI}>
      WHERE {
        ?s ns:name ?name .
        ?s ns:price ?price .
      }
      ORDER BY ASC(?name)
    `);

    expect(books).toMatchSnapshot();
  });

  it('Clears a graph', async () => {
    await db.query(`CLEAR GRAPH <${TEST_GRAPH_IRI}>`);
    const [{ count }] = await db.query(
      `SELECT COUNT(*) as ?count FROM <${TEST_GRAPH_IRI}> WHERE { ?s ?p ?o }`,
    );
    await expect(count.value).toBe('0');
  });
});
