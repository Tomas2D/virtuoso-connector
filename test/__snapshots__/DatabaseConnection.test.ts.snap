// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Database connection with queries Retrieves data 1`] = `
Array [
  Object {
    "name": Literal {
      "datatype": NamedNode {
        "termType": "NamedNode",
        "value": "http://www.w3.org/1999/02/22-rdf-syntax-ns#langString",
      },
      "language": "en",
      "termType": "Literal",
      "value": "Book #1",
    },
    "price": Literal {
      "datatype": NamedNode {
        "termType": "NamedNode",
        "value": "http://www.w3.org/2001/XMLSchema#integer",
      },
      "language": "",
      "termType": "Literal",
      "value": "100",
    },
  },
  Object {
    "name": Literal {
      "datatype": NamedNode {
        "termType": "NamedNode",
        "value": "http://www.w3.org/1999/02/22-rdf-syntax-ns#langString",
      },
      "language": "en",
      "termType": "Literal",
      "value": "Book #1",
    },
    "price": Literal {
      "datatype": NamedNode {
        "termType": "NamedNode",
        "value": "http://www.w3.org/2001/XMLSchema#integer",
      },
      "language": "",
      "termType": "Literal",
      "value": "200",
    },
  },
  Object {
    "name": Literal {
      "datatype": NamedNode {
        "termType": "NamedNode",
        "value": "http://www.w3.org/1999/02/22-rdf-syntax-ns#langString",
      },
      "language": "en",
      "termType": "Literal",
      "value": "Book #2",
    },
    "price": Literal {
      "datatype": NamedNode {
        "termType": "NamedNode",
        "value": "http://www.w3.org/2001/XMLSchema#integer",
      },
      "language": "",
      "termType": "Literal",
      "value": "100",
    },
  },
  Object {
    "name": Literal {
      "datatype": NamedNode {
        "termType": "NamedNode",
        "value": "http://www.w3.org/1999/02/22-rdf-syntax-ns#langString",
      },
      "language": "en",
      "termType": "Literal",
      "value": "Book #2",
    },
    "price": Literal {
      "datatype": NamedNode {
        "termType": "NamedNode",
        "value": "http://www.w3.org/2001/XMLSchema#integer",
      },
      "language": "",
      "termType": "Literal",
      "value": "200",
    },
  },
]
`;
